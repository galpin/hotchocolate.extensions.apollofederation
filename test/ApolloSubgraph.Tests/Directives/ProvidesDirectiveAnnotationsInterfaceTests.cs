using System.Collections.Generic;
using System.Threading.Tasks;
using HotChocolate.Types;
using Microsoft.Extensions.DependencyInjection;
using Xunit;
using static HotChocolate.Extensions.ApolloSubgraph.Test;

namespace HotChocolate.Extensions.ApolloSubgraph.Directives;

public class ProvidesDirectiveAnnotationsInterfaceTests
{
    [Fact]
    public async Task When_provides_is_specified_on_interface()
    {
        var schema = await BuildSchemaAsync(builder =>
        {
            builder.AddType<IReview>();
            builder.AddType<Review>();
            builder.AddType<Product>();
            builder.AddQueryType();
        });

        var sut = schema.GetType<InterfaceType>(nameof(IReview));

        Assert.Collection(
            sut.Fields["products"].Directives,
            x => AssertEx.Directive(x, "provides", ("fields", "\"name\"")));
        await schema.QuerySdlAndMatchSnapshotAsync();
    }

    [GraphQLExtends]
    public interface IReview
    {
        [GraphQLKey]
        [GraphQLExternal]
        int Id { get; }

        [GraphQLProvides("name")]
        IReadOnlyList<Product> Products { get; }
    }

    public class Review : IReview
    {
        public Review(int id, IReadOnlyList<Product> products)
        {
            Id = id;
            Products = products;
        }

        [GraphQLKey]
        [GraphQLExternal]
        public int Id { get; }

        [GraphQLProvides("name")]
        public IReadOnlyList<Product> Products { get; }
    }

    [GraphQLExtends]
    public class Product
    {
        public Product(string upc)
        {
            Upc = upc;
        }

        [GraphQLKey]
        public string Upc { get; }
    }
}