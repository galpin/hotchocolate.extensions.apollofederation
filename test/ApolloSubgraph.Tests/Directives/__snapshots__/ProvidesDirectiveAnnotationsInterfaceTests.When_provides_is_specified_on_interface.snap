{
  "data": {
    "_service": {
      "sdl": "interface IReview @extends @key(fields: \u0022id\u0022) {\n  id: Int! @external\n  products: [Product!]! @provides(fields: \u0022name\u0022)\n}\n\ntype Product @extends @key(fields: \u0022upc\u0022) {\n  upc: String!\n}\n\ntype Review implements IReview @key(fields: \u0022id\u0022) {\n  id: Int! @external\n  products: [Product!]! @provides(fields: \u0022name\u0022)\n}"
    }
  }
}
